### PR TITLE
 [FLINK-36240][runtime/metrics] Fix incorrect Ports display in the PrometheusReporter constructor in case of the httpServer creation failure

### PR DESCRIPTION
**Pull request by eon2208**
## What is the purpose of the change
* Fixes incorrect Ports display in the PrometheusReporter constructor in case of the httpServer creation failure. Instead of printing the port numbers, the error message currently displays an Iterator object reference, which is not useful for debugging.

## Brief change log
* Moved the `PortsRange` from the runtime module to the core module
* Use `PortsRange` utility class from the core module instead of the `iterator`. This ensures that the actual ports are printed in the exception message instead of the Iterator reference.
* Update the tests to rely one on the `PortsRange` as the `PrometheusReposter` only constructor parameter instead of the Iterator

## Verifying this change
 - Extends the existing cannotStartTwoReportersOnSamePort test by checking the exception message thrown to verify that the log message contains the actual ports and not the object reference.
- Examples 
   - before: `Could not start PrometheusReporter HTTP server on any configured port. Ports: org.apache.flink.util.UnionIterator@67065fd3` 
   - after: `Could not start PrometheusReporter HTTP server on any configured port. Ports: [9000, 9001, 9002, ...]` 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no 
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### TBD
I have also considered changing the constructor parameter from `Iterable<Integer>` to `List<Integer>`, as it would provide a more straightforward way to manage the port numbers. This change would require minor adjustments in the codebase, namely modifying the `PrometheusReporterFactory` and some test cases, which currently use `Iterable<Integer>` and would need conversion to `List<Integer>`. I’d like to discuss if this change is worthwhile, as it could improve code clarity but requires slightly more effort.

Alternatively, converting the `Iterable<Integer>` to a `List<Integer>` at the beginning of the constructor, and then attempting to create the HTTP server using this list, could also be a solution. In my opinion, this would make the code a bit cleaner and easier to read. However, it would require iterating over the ports twice: once when converting the `Iterable` to a list, and again during the actual HTTP server initialization. While this extra iteration may have a small impact on performance, I don't think it would be significant enough to cause issues.

**Questions for Discussion**:
- Is it worth refactoring the constructor to use `List<Integer>` instead of `Iterable<Integer>`, considering the minor adjustments required across the codebase?
- Would the conversion from `Iterable` to `List` in the constructor be a good trade-off for improved readability, even if it introduces a second iteration?
- Does the current implementation fit well as-is, or could these refinements be valuable improvements?

### CODE REVIEW

**Comment by ferenc-csaky** [comment]
One other alternative would be to initialize a `PortRange` in the factory and pass that object in the `PrometheusReporter` ctor. `PortRange` stores the original original string config, and is converted into that in case of `toString()` so it will display the ports correctly. This way we do not need to care about converting to a list at all, simply getting the iterator through the `PortRange`, which was my original intent when I wrote that util. :)

A con of this approach might be the usage of a fairly specific class instead of a common interface in the `PrometheusReporter` ctor, but since that ctor is not meant to use and should be instantiated by the factory IMO it does not really matter.

WDYT?

**Comment by eon2208** [comment]
Using the PortRange as a domain wrapper for the ports seems like a good choice. The question is whether adding the extra dependency to the `flink-metrics-prometheus` module for the single class is worth it (IMO it makes the code more intuitive to read). If so, then I would go for it.

**Comment by ferenc-csaky** [comment]
> Using the PortRange as a domain wrapper for the ports seems like a good choice. The question is whether adding the extra dependency to the `flink-metrics-prometheus` module for the single class is worth it (IMO it makes the code more intuitive to read). If so, then I would go for it.

You make a good paint here, did not check the code in the morning, I thought `PortRange` lives in the same module as `NetUtils`, but it seems I put it into `flink-runtime` as I added the improvement in that module back in the day. 

`PortRange` is  general utility class that makes the `NetUtils#getPortRangeFromString` a bit more user-friendly so I think it makes sense to move them into the same module.

@gaborgsomogyi if you have some time, I would appreciate your opinion on this. If agreed, I'll open another ticket about the utility move.

**Comment by eon2208** [comment]
> PortRange is general utility class that makes the NetUtils#getPortRangeFromString a bit more user-friendly so I think it makes sense to move them into the same module.

Strongly agree, if the ticket would be created and you have something more urgent to do, I will volunteer to do it as well.

**Comment by ferenc-csaky** [comment]
> > PortRange is general utility class that makes the NetUtils#getPortRangeFromString a bit more user-friendly so I think it makes sense to move them into the same module.
> 
> Strongly agree, if the ticket would be created and you have something more urgent to do, I will volunteer to do it as well.

Thank you, if we go into that route, I will create the ticket, and assign it to you! I think it can be solved in this one pull-request, just create 2 separate commits for the 2 change, and we can "rebase and merge".

**Comment by ferenc-csaky** [comment]
cc @mateczagany

**Comment by ferenc-csaky** [comment]
> I have looked into the `PortRange` class and I think using that would solve the problem this JIRA is trying to tackle. I like the solution, so just to confirm we would have 2 commits:
> 
> * Move `PortRange` class from flink-runtime to flink-core module, so no additional dependencies required for flink-metrics-prometheus.
> * Change constructor of `PrometheusReporter` to require a single parameter of `PortRange`.

Yep, exactly what you described.

**Comment by ferenc-csaky** [comment]
Created FLINK-36316 to track the util move, @eon2208 feel free to proceed with the changes and force-push this PR with the 2 separate commits.

**Comment by eon2208** [comment]
@ferenc-csaky Done!

**Comment by ferenc-csaky** [comment]
Logic LGTM, lets wait for the CI run, will approve when it is successful.

**Comment by ferenc-csaky** [comment]
FYI: compile failed because of code style, make sure to run `mvn spotless:apply` to format your code.

**Comment by eon2208** [comment]
Thanks for the reminder.I changed one thing after the spotless apply and auroformatting made a mess of it (Have seen the docs on setting up the IDE and will do it later).I have double checked it now and am pushed properly formatted code.

**Review by ferenc-csaky — APPROVED** [review]
LGTM, will merge by tomorrow EOD CEST if no objections.

**Review by mateczagany — APPROVED** [review]

**Pull request merged**
